File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.js
function solution(gems) {
    // Array는 .length 로 길이
    const N = gems.length;
    // Set는 .size 로 길이
    const M = new Set(gems).size;
    
    const gem_dict = new Map();
    
    let result_s = 0;
    let result_e = N - 1;
    
    let start = 0;
    
    for (let end = 0; end < N; end++) {
        // end 포인터의 보석 이름
        const gem = gems[end];
        // map.set(key, value): 값 넣기
        // map.get(key): 값 읽기
        // -> gem 이 있다면 기존 값을 +1,
        // -> gem이 없다면 0 + 1
        gem_dict.set(gem, (gem_dict.get(gem) || 0) + 1);
        
        while (gem_dict.size === M) {
            if (end - start < result_e - result_s) {
                result_e = end;
                result_s = start;
            }
            
            const startGem = gems[start];
            gem_dict.set(startGem, gem_dict.get(startGem) - 1);
            if (gem_dict.get(startGem) === 0) {
                gem_dict.delete(startGem);
            }
            start += 1;
        }
    }
    return [result_s + 1, result_e + 1];
}

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
from collections import defaultdict

def solution(gems):
    N = len(gems)
    M = len(set(gems))
    
    gem_dict = defaultdict(int)
    
    result_s = 0
    result_e = N-1
    
    start = 0
    
    # end 포인터를 하나씩 뒤로 민다 (추가 개념)
    for end in range(N):
        gem_dict[gems[end]] += 1
        
        while len(gem_dict) == M:
            if end - start < result_e - result_s:
                result_e = end
                result_s = start
                
            # 최소를 찾기 위해,
            # start 포인터를 하나씩 뒤로 민다 (제외 개념)
            gem_dict[gems[start]] -= 1
            if gem_dict[gems[start]] == 0:
                del gem_dict[gems[start]]
            start += 1
    
    return [result_s + 1, result_e + 1]